test/verify/machineslib.py

```diff
@@ -790,12 +790,17 @@ def testVCPU(self):
 
         b.wait(lambda: m.execute("virsh dumpxml subVmTest1 | xmllint --xpath '/domain/cpu/topology[@sockets=\"2\"][@threads=\"1\"][@cores=\"2\"]' -"))
 
+        # Run VM - this ensures that the internal state is updated before we move on.
+        # We need this here because we can't wait for UI updates after we open the modal dialog.
+        b.click("#vm-subVmTest1-run")
+        b.wait_in_text("#vm-subVmTest1-state", "running")
+
         # Open dialog
         b.wait_present("#vm-subVmTest1-vcpus-count")
         b.click("#vm-subVmTest1-vcpus-count")
 
         b.wait_present(".vcpu-detail-modal-table")
-        b.wait_not_present(".machines-vcpu-caution")
+        b.wait_present(".machines-vcpu-caution")
```

Comment:
So the ```virsh dumpxml``` check here
https://github.com/cockpit-project/cockpit/blob/d9067fda912cb7a877f5d740f4b1a748e237a502/test/verify/machineslib.py#L808
is completely useless because we checked the xml values here
https://github.com/cockpit-project/cockpit/blob/d9067fda912cb7a877f5d740f4b1a748e237a502/test/verify/machineslib.py#L791
and afterwards we never clicked apply, thus we caused no changes.

The state update for changed VCPU settings is a reaction to a signal caused by redefining the domain.
What we do:
(Following applies for virsh provider where test fails)
1) Apply new VCPU settings from UI
2) Internally call ```virsh define``` to update  the domain XML
3) We receive a signal in virsh event loop, that a domain got re-defined
4) As a reaction to this signal we dispatch an action that will update the state

So, coming to the test:
1) t is clicking apply
2) then waits for the ```virsh dumpxml``` to show correct values
4) at this point it means that we are done with calling ```virsh define```. (Previous step number 2)
The signal will follow, of course there should not be great delay,  and trigger the state update.
5) I am afraid that the test clicks here to open the modal before the state change actually got triggered (previous number 4), and I think we should not rely on how fast some call finished inside the test.

Your suggestion in the last paragraph will work as well, but I think its more confusing.

And regarding the caution check, I could move it earlier, where the domain is not running either.

And regarding if the state is not computed at dialog open time, at that point we just read what is the current state, so if the state update gets triggered right after we open the dialog, we have non updated state.